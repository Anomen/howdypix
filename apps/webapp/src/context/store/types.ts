import React from "react";

export type StoreContextData = {
  currentAlbum: string | null;
  currentSource: string | null;
  withLayout: boolean;
  rightPanel: React.ReactElement | null;
  setCurrentAlbum: (album: StoreContextData["currentAlbum"]) => void;
  setCurrentSource: (source: StoreContextData["currentSource"]) => void;
  setRightPanel: (panel: StoreContextData["rightPanel"]) => void;
  setWithLayout: (panel: StoreContextData["withLayout"]) => void;
};
