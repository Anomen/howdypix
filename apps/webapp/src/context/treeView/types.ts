import {
  GetTreeAlbums,
  GetTreeSources,
} from "../../__generated__/schema-types";

export type TreeItem = {
  album: string | null;
  source: string;
};

export type TreeItemWithParent = TreeItem & {
  parent: string | null;
};

export type AlbumWithNodeId = GetTreeAlbums & {
  nodeId: string;
};
export type SourceWithNodeId = GetTreeSources & {
  nodeId: string;
};

export type TreeViewContextData = {
  expand: (item: TreeItem) => void;
  collapse: (item: TreeItemWithParent) => void;
  toggle: (item: TreeItemWithParent) => void;
  expandedNodeIds: string[];
  sources: SourceWithNodeId[];
  albums: AlbumWithNodeId[];
};
