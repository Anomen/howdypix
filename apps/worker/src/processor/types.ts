import { ExifData, HFile } from "@howdypix/shared-types";

export type FetchExif = (root: string, path: string) => Promise<ExifData>;
export type CreateThumbnails = (
  thumbnailsDir: string,
  root: string,
  hfile: HFile
) => Promise<string[]>;

export type FileProcessor = {
  fetchExif: FetchExif;
  createThumbnails: CreateThumbnails;
};
