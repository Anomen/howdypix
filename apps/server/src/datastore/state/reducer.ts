import { Reducer } from "redux";
import { uniqWith, isEqual } from "lodash";

type StateSearch = {
  lastUpdatedPhoto: Date | null;
  album: string;
  source: string;
};

type StateAlbum = {
  lastUpdatedPhoto: Date | null;
};

type StateSearchAlbumRelation = {
  searchId: number;
  albumId: number;
};

export type State = {
  searches: {
    [id: number]: StateSearch;
  };
  albums: {
    [id: number]: StateAlbum;
  };
  searchAlbumRelation: StateSearchAlbumRelation[];
};

export type Action =
  | {
      type: "RESET_STORE";
      data: State;
    }
  | {
      type: "UPDATE_SEARCH_LAST_UPDATED_PHOTO";
      searchId: number;
      data: {
        lastUpdatedPhoto: Date;
      };
    }
  | {
      type: "STORE_NEW_SEARCH";
      searchId: number;
      data: StateSearch;
    }
  | {
      type: "UPDATE_ALBUM_LAST_UPDATED_PHOTO";
      albumId: number;
      data: {
        lastUpdatedPhoto: Date;
      };
    }
  | {
      type: "STORE_NEW_ALBUM";
      albumId: number;
      data: StateAlbum;
    }
  | {
      type: "STORE_NEW_SEARCH_ALBUM_RELATION";
      albumId: number;
      searchId: number;
    };

// ========================================================================
// Reducer
// ========================================================================
export const reducer: Reducer<State, Action> = (
  state = { searches: {}, albums: {}, searchAlbumRelation: [] },
  action
) => {
  // console.log(action);
  switch (action.type) {
    case "RESET_STORE":
      return action.data;

    case "UPDATE_SEARCH_LAST_UPDATED_PHOTO": {
      const lastUpdatedPhoto =
        state.searches[action.searchId]?.lastUpdatedPhoto;

      if (
        (lastUpdatedPhoto && action.data.lastUpdatedPhoto > lastUpdatedPhoto) ||
        lastUpdatedPhoto === null
      ) {
        return {
          ...state,
          searches: {
            ...state.searches,
            [action.searchId]: {
              ...state.searches[action.searchId],
              lastUpdatedPhoto: new Date(action.data.lastUpdatedPhoto),
            },
          },
        };
      }

      return state;
    }

    case "UPDATE_ALBUM_LAST_UPDATED_PHOTO": {
      const lastUpdatedPhoto = state.albums[action.albumId]?.lastUpdatedPhoto;

      if (
        (lastUpdatedPhoto && action.data.lastUpdatedPhoto > lastUpdatedPhoto) ||
        lastUpdatedPhoto === null
      ) {
        return {
          ...state,
          albums: {
            ...state.albums,
            [action.albumId]: {
              lastUpdatedPhoto: new Date(action.data.lastUpdatedPhoto),
            },
          },
        };
      }

      return state;
    }

    case "STORE_NEW_ALBUM":
      return {
        ...state,
        albums: {
          ...state.albums,
          [action.albumId]: {
            ...action.data,
            lastUpdatedPhoto: action.data.lastUpdatedPhoto
              ? new Date(action.data.lastUpdatedPhoto)
              : null,
          },
        },
      };

    case "STORE_NEW_SEARCH":
      return {
        ...state,
        searches: {
          ...state.searches,
          [action.searchId]: {
            ...action.data,
            lastUpdatedPhoto: action.data.lastUpdatedPhoto
              ? new Date(action.data.lastUpdatedPhoto)
              : null,
          },
        },
        searchAlbumRelation: uniqWith(
          [
            ...state.searchAlbumRelation,
            {
              searchId: action.searchId,
              albumId: action.data.album,
            },
          ],
          isEqual
        ),
      };

    case "STORE_NEW_SEARCH_ALBUM_RELATION":
      return {
        ...state,
        searchAlbumRelation: [
          ...state.searchAlbumRelation,
          ...(state.searchAlbumRelation.find(
            (relation) =>
              relation.albumId === action.albumId &&
              relation.searchId === action.searchId
          )
            ? [{ albumId: action.albumId, searchId: action.searchId }]
            : []),
        ],
      };

    default:
      return state;
  }
};
