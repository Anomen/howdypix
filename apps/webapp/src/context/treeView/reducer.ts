import { uniq, uniqBy } from "lodash";
import { QueryLazyOptions } from "@apollo/react-hooks";
import {
  GetTreeAlbums,
  GetTreeQuery,
  GetTreeQueryVariables,
  GetTreeSources,
} from "../../__generated__/schema-types";
import {
  AlbumWithNodeId,
  SourceWithNodeId,
  TreeItem,
  TreeItemWithParent,
} from "./types";

// ========================================================================
// Utility functions
// ========================================================================
const selectParents = (
  data: GetTreeQuery["getTree"]["albums"] | null | undefined,
  dir: string
): { [key: string]: boolean } => {
  let ret: { [key: string]: boolean } = {};

  if (!data) {
    ret[dir] = true;
  } else {
    data.forEach((d): void => {
      if (d.dir === dir) {
        ret[d.dir] = true;

        if (d.parentDir) {
          ret = {
            ...ret,
            ...selectParents(data, d.parentDir),
          };
        }
      }
    });
  }

  return ret;
};

const calculateExpandedNodeIds = (
  data: AlbumWithNodeId[],
  selectedItems: { [name: string]: boolean }
): string[] => {
  const ret: string[] = [];

  data.forEach((d): void => {
    if (d.dir && selectedItems[d.dir]) {
      ret.push(d.nodeId);
    }
    if (d.parentDir === "" && selectedItems[d.source]) {
      ret.push(d.source);
    }
  });

  return uniq(ret);
};

// ========================================================================
// Types
// ========================================================================
export type State = {
  fetchedAlbums: AlbumWithNodeId[];
  fetchedSources: SourceWithNodeId[];
  visibleLeaves: {
    [key: string]: boolean;
  };
  expandedNodeIds: string[];
};

export type Actions =
  | {
      type: "DATA_FETCHED";
      albums: GetTreeAlbums[];
      sources: GetTreeSources[];
      variables: {
        album: string;
        source: string;
      };
    }
  | {
      type: "DISPLAY_LEAF";
      album: string;
      source: string;
    }
  | {
      type: "UPDATE_EXPANDED_NODE_IDS";
    }
  | (TreeItem & {
      type: "EXPAND";
      fetchTree: (
        options?: QueryLazyOptions<GetTreeQueryVariables> | undefined
      ) => void;
    })
  | (TreeItemWithParent & {
      type: "COLLAPSE";
    })
  | (TreeItemWithParent & {
      type: "TOGGLE";
      fetchTree: (
        options?: QueryLazyOptions<GetTreeQueryVariables> | undefined
      ) => void;
    });

// ========================================================================
// Reducer
// ========================================================================
export const reducer = (state: State, action: Actions): State => {
  let newState: State = state;

  switch (action.type) {
    case "DATA_FETCHED":
      newState = {
        ...newState,
        fetchedSources: uniqBy(
          [...(newState.fetchedSources ?? []), ...action.sources],
          "name"
        ).map(
          (source): SourceWithNodeId => ({
            ...source,
            nodeId: source.name,
          })
        ),
        fetchedAlbums: uniqBy(
          [...(newState.fetchedAlbums ?? []), ...action.albums],
          "dir"
        ).map((album) => ({
          ...album,
          nodeId: album.source + album.dir,
        })),
      };

      return reducer(newState, {
        type: "DISPLAY_LEAF",
        source: action.variables.source,
        album: action.variables.album,
      });

    case "DISPLAY_LEAF":
      return reducer(
        {
          ...newState,
          visibleLeaves: {
            ...newState.visibleLeaves,
            ...(action.album !== ""
              ? selectParents(newState.fetchedAlbums, action.album)
              : {}),
            [action.source]: true,
          },
        },
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "UPDATE_EXPANDED_NODE_IDS":
      return {
        ...newState,
        expandedNodeIds: calculateExpandedNodeIds(
          newState.fetchedAlbums,
          newState.visibleLeaves
        ),
      };

    case "EXPAND":
      if (action.album === null) {
        if (newState.visibleLeaves[action.source] === undefined) {
          action.fetchTree({
            variables: {
              album: "",
              source: action.source,
            },
          });
        }
      } else if (newState.visibleLeaves[action.album] === undefined) {
        action.fetchTree({
          variables: {
            album: action.album ?? "",
            source: action.source,
          },
        });
      }

      // different than original
      return reducer(
        reducer(newState, {
          type: "DISPLAY_LEAF",
          source: action.source,
          album: action.album ?? "",
        }),
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "COLLAPSE":
      return reducer(
        {
          ...newState,
          visibleLeaves: {
            ...newState.visibleLeaves,
            ...(action.album !== null
              ? { [action.album]: false }
              : { [action.source]: false }),
            ...(action.parent ? { [action.parent]: true } : {}),
          },
        },
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "TOGGLE":
      if (action.album !== null) {
        if (newState.visibleLeaves[action.album]) {
          return reducer(newState, {
            ...action,
            type: "COLLAPSE",
          });
        }

        return reducer(newState, {
          ...action,
          type: "EXPAND",
        });
      }

      if (newState.visibleLeaves[action.source]) {
        return reducer(newState, {
          ...action,
          type: "COLLAPSE",
        });
      }

      return reducer(newState, {
        ...action,
        type: "EXPAND",
      });

    default:
      return newState;
  }
};
