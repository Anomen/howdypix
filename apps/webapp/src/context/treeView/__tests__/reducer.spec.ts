import { reducer } from "../reducer";
import { AlbumWithNodeId, SourceWithNodeId } from "../types";

const createFakeAlbum = (
  id: number,
  sourceId: number,
  parentDirId: number | null = null,
  overrides: Partial<AlbumWithNodeId> = {}
): AlbumWithNodeId => ({
  nodeId: `album${id}`,
  dir: `album${id}`,
  parentDir: parentDirId ? `album${parentDirId}` : "",
  source: `source${sourceId}`,
  nbAlbums: 0,
  nbImages: id,
  preview: `preview${id}`,
  ...overrides,
});

const createFakeSource = (
  id: number,
  overrides: Partial<SourceWithNodeId> = {}
): SourceWithNodeId => ({
  nodeId: `source${id}`,
  name: `source${id}`,
  nbAlbums: 0,
  nbImages: id,
  preview: `preview${id}`,
  ...overrides,
});

describe("treeView reducer", () => {
  describe("for the action DATA_FETCHED", () => {
    test("must merge the fetched information in the state", async () => {
      const result = reducer(
        {
          visibleLeaves: {},
          fetchedAlbums: [createFakeAlbum(1, 2)],
          expandedNodeIds: [],
          fetchedSources: [
            createFakeSource(1, { nbAlbums: 1 }),
            createFakeSource(2, { nbAlbums: 1 }),
          ],
        },
        {
          type: "DATA_FETCHED",
          variables: { album: "album1", source: "source1" },
          sources: [
            createFakeSource(1, { nbAlbums: 1 }),
            createFakeSource(2, { nbAlbums: 1 }),
          ],
          albums: [createFakeAlbum(2, 2)],
        }
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe("for the action DISPLAY_LEAF", () => {
    test("must update the visible leaves state and the expandedNodeIds for an album", async () => {
      const { visibleLeaves, expandedNodeIds } = reducer(
        {
          visibleLeaves: {},
          fetchedAlbums: [
            createFakeAlbum(1, 1),
            createFakeAlbum(2, 1),
            createFakeAlbum(3, 1, null, { nbAlbums: 1 }),
            createFakeAlbum(4, 1, 3, { nbAlbums: 1 }),
            createFakeAlbum(5, 1, 4),
          ],
          expandedNodeIds: [],
          fetchedSources: [
            createFakeSource(1, { nbAlbums: 1 }),
            createFakeSource(2),
          ],
        },
        {
          type: "DISPLAY_LEAF",
          album: "album5",
          source: "source1",
        }
      );

      expect(visibleLeaves).toMatchSnapshot();
      expect(expandedNodeIds).toMatchSnapshot();
    });

    test("must update the visible leaves state and the expandedNodeIds for a source", async () => {
      const { visibleLeaves, expandedNodeIds } = reducer(
        {
          visibleLeaves: {},
          fetchedAlbums: [
            createFakeAlbum(1, 1),
            createFakeAlbum(2, 1),
            createFakeAlbum(3, 1, null, { nbAlbums: 1 }),
            createFakeAlbum(4, 1, 3, { nbAlbums: 1 }),
            createFakeAlbum(5, 1, 4),
          ],
          expandedNodeIds: [],
          fetchedSources: [
            createFakeSource(1, { nbAlbums: 3 }),
            createFakeSource(2),
          ],
        },
        {
          type: "DISPLAY_LEAF",
          album: "",
          source: "source1",
        }
      );

      expect(visibleLeaves).toMatchSnapshot();
      expect(expandedNodeIds).toMatchSnapshot();
    });
  });

  describe("for the action UPDATE_EXPANDED_NODE_IDS", () => {
    test("must update the expandedNodeIds", async () => {
      const { expandedNodeIds } = reducer(
        {
          visibleLeaves: {
            album1: true,
            album2: true,
            album3: true,
            source1: true,
            source2: true,
          },
          fetchedAlbums: [
            createFakeAlbum(1, 1),
            createFakeAlbum(2, 1),
            createFakeAlbum(3, 2),
          ],
          expandedNodeIds: [],
          fetchedSources: [
            createFakeSource(1, { nbAlbums: 2 }),
            createFakeSource(2, { nbAlbums: 1 }),
          ],
        },
        {
          type: "UPDATE_EXPANDED_NODE_IDS",
        }
      );

      expect(expandedNodeIds).toMatchSnapshot();
    });

    test("must update the expandedNodeIds when everything is false", async () => {
      const { expandedNodeIds } = reducer(
        {
          visibleLeaves: {
            album1: false,
            album2: false,
            album3: false,
            source1: false,
            source2: false,
          },
          fetchedAlbums: [
            createFakeAlbum(1, 1),
            createFakeAlbum(2, 1),
            createFakeAlbum(3, 2),
          ],
          expandedNodeIds: ["album1", "source1", "album2", "album3", "source2"],
          fetchedSources: [
            createFakeSource(1, { nbAlbums: 2 }),
            createFakeSource(2, { nbAlbums: 1 }),
          ],
        },
        {
          type: "UPDATE_EXPANDED_NODE_IDS",
        }
      );

      expect(expandedNodeIds).toMatchSnapshot();
    });
  });

  describe("for the action EXPAND", () => {
    test("must fetch the album if not fetched before", async () => {
      const fetchTree = jest.fn();

      reducer(
        {
          visibleLeaves: {},
          fetchedAlbums: [],
          expandedNodeIds: [],
          fetchedSources: [createFakeSource(1, { nbAlbums: 2 })],
        },
        {
          type: "EXPAND",
          source: "source1",
          album: "album1",
          fetchTree,
        }
      );

      expect(fetchTree).toHaveBeenCalledWith({
        variables: {
          album: "album1",
          source: "source1",
        },
      });
    });

    test("must fetch the source if not fetched before", async () => {
      const fetchTree = jest.fn();

      reducer(
        {
          visibleLeaves: {},
          fetchedAlbums: [],
          expandedNodeIds: [],
          fetchedSources: [createFakeSource(1, { nbAlbums: 2 })],
        },
        {
          type: "EXPAND",
          source: "source1",
          album: null,
          fetchTree,
        }
      );

      expect(fetchTree).toHaveBeenCalledWith({
        variables: {
          album: "",
          source: "source1",
        },
      });
    });

    test("must display the correct leaf if it's a source", async () => {
      const fetchTree = jest.fn();

      const { visibleLeaves } = reducer(
        {
          visibleLeaves: { source1: false, album1: false },
          fetchedAlbums: [
            createFakeAlbum(1, 1, null, { nbAlbums: 1 }),
            createFakeAlbum(1, 1, 2),
          ],
          expandedNodeIds: [],
          fetchedSources: [createFakeSource(1, { nbAlbums: 2 })],
        },
        {
          type: "EXPAND",
          source: "source1",
          album: null,
          fetchTree,
        }
      );

      expect(fetchTree).not.toHaveBeenCalled();
      expect(visibleLeaves).toMatchSnapshot();
    });

    test("must display the correct leaf if it's an album", async () => {
      const fetchTree = jest.fn();

      const { visibleLeaves } = reducer(
        {
          visibleLeaves: { source1: true, album1: false, album2: true },
          fetchedAlbums: [createFakeAlbum(1, 1)],
          expandedNodeIds: [],
          fetchedSources: [createFakeSource(1, { nbAlbums: 2 })],
        },
        {
          type: "EXPAND",
          source: "source1",
          album: "album1",
          fetchTree,
        }
      );

      expect(fetchTree).not.toHaveBeenCalled();
      expect(visibleLeaves).toMatchSnapshot();
    });
  });

  describe("for the action COLLAPSE", () => {
    test("must hide the corresponding leaves if hidding an album", async () => {
      const { visibleLeaves, expandedNodeIds } = reducer(
        {
          visibleLeaves: {
            album1: true,
            album2: true,
            album3: true,
            source1: true,
            source2: true,
          },
          fetchedAlbums: [
            createFakeAlbum(1, 1),
            createFakeAlbum(2, 1),
            createFakeAlbum(3, 2),
          ],
          expandedNodeIds: ["album1", "source1", "album2", "album3", "source2"],
          fetchedSources: [
            createFakeSource(1, { nbAlbums: 2 }),
            createFakeSource(2, { nbAlbums: 1 }),
          ],
        },
        {
          type: "COLLAPSE",
          album: "album1",
          source: "source1",
          parent: null,
        }
      );

      expect(visibleLeaves).toMatchSnapshot();
      expect(expandedNodeIds).toMatchSnapshot();
    });

    test("must hide the corresponding leaves if hiding a source", async () => {
      const { visibleLeaves, expandedNodeIds } = reducer(
        {
          visibleLeaves: {
            album1: true,
            album2: true,
            album3: true,
            source1: true,
            source2: true,
          },
          fetchedAlbums: [
            createFakeAlbum(1, 1),
            createFakeAlbum(2, 1),
            createFakeAlbum(3, 2),
          ],
          expandedNodeIds: ["album1", "source1", "album2", "album3", "source2"],
          fetchedSources: [
            createFakeSource(1, { nbAlbums: 2 }),
            createFakeSource(2, { nbAlbums: 1 }),
          ],
        },
        {
          type: "COLLAPSE",
          album: null,
          source: "source1",
          parent: null,
        }
      );

      expect(visibleLeaves).toMatchSnapshot();
      expect(expandedNodeIds).toMatchSnapshot();
    });
  });

  describe("for the action TOGGLE", () => {
    const base = {
      visibleLeaves: {
        album1: true,
        album2: true,
        album3: true,
        source1: true,
        source2: true,
      },
      fetchedAlbums: [
        createFakeAlbum(1, 1),
        createFakeAlbum(2, 1),
        createFakeAlbum(3, 2),
      ],
      expandedNodeIds: ["album1", "source1", "album2", "album3", "source2"],
      fetchedSources: [
        createFakeSource(1, { nbAlbums: 2 }),
        createFakeSource(2, { nbAlbums: 1 }),
      ],
    };

    test("must call collapse if it's visible", async () => {
      const fetchTree = jest.fn();

      const { visibleLeaves, expandedNodeIds } = reducer(base, {
        type: "TOGGLE",
        album: "album1",
        source: "source1",
        parent: null,
        fetchTree,
      });

      expect(fetchTree).not.toHaveBeenCalled();
      expect(visibleLeaves).toMatchSnapshot();
      expect(expandedNodeIds).toMatchSnapshot();
    });

    test("must call expand if it's hidden", async () => {
      const fetchTree = jest.fn();

      const { visibleLeaves, expandedNodeIds } = reducer(
        {
          ...base,
          ...{
            visibleLeaves: {
              ...base.visibleLeaves,
              ...{ album1: false, album2: false },
            },
            expandedNodeIds: ["source1", "album3", "source2"],
          },
        },
        {
          type: "TOGGLE",
          album: "album1",
          source: "source1",
          parent: null,
          fetchTree,
        }
      );

      expect(fetchTree).not.toHaveBeenCalled();
      expect(visibleLeaves).toMatchSnapshot();
      expect(expandedNodeIds).toMatchSnapshot();
    });

    test("must fetch if it's not been fetched before", async () => {
      const fetchTree = jest.fn();

      reducer(base, {
        type: "TOGGLE",
        album: "album5",
        source: "source2",
        parent: null,
        fetchTree,
      });

      expect(fetchTree).toHaveBeenCalledWith({
        variables: {
          album: "album5",
          source: "source2",
        },
      });
    });
  });
});
