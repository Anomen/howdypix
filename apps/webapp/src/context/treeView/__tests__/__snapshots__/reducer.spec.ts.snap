// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treeView reducer for the action COLLAPSE must hide the corresponding leaves if hidding an album 1`] = `
Object {
  "album1": false,
  "album2": true,
  "album3": true,
  "source1": true,
  "source2": true,
}
`;

exports[`treeView reducer for the action COLLAPSE must hide the corresponding leaves if hidding an album 2`] = `
Array [
  "source1",
  "album2",
  "album3",
  "source2",
]
`;

exports[`treeView reducer for the action COLLAPSE must hide the corresponding leaves if hiding a source 1`] = `
Object {
  "album1": true,
  "album2": true,
  "album3": true,
  "source1": false,
  "source2": true,
}
`;

exports[`treeView reducer for the action COLLAPSE must hide the corresponding leaves if hiding a source 2`] = `
Array [
  "album1",
  "album2",
  "album3",
  "source2",
]
`;

exports[`treeView reducer for the action DATA_FETCHED must merge the fetched information in the state 1`] = `
Object {
  "expandedNodeIds": Array [
    "source2album1",
  ],
  "fetchedAlbums": Array [
    Object {
      "dir": "album1",
      "nbAlbums": 0,
      "nbImages": 1,
      "nodeId": "source2album1",
      "parentDir": "",
      "preview": "preview1",
      "source": "source2",
    },
    Object {
      "dir": "album2",
      "nbAlbums": 0,
      "nbImages": 2,
      "nodeId": "source2album2",
      "parentDir": "",
      "preview": "preview2",
      "source": "source2",
    },
  ],
  "fetchedSources": Array [
    Object {
      "name": "source1",
      "nbAlbums": 1,
      "nbImages": 1,
      "nodeId": "source1",
      "preview": "preview1",
    },
    Object {
      "name": "source2",
      "nbAlbums": 1,
      "nbImages": 2,
      "nodeId": "source2",
      "preview": "preview2",
    },
  ],
  "visibleLeaves": Object {
    "album1": true,
    "source1": true,
  },
}
`;

exports[`treeView reducer for the action DISPLAY_LEAF must update the visible leaves state and the expandedNodeIds for a source 1`] = `
Object {
  "source1": true,
}
`;

exports[`treeView reducer for the action DISPLAY_LEAF must update the visible leaves state and the expandedNodeIds for a source 2`] = `
Array [
  "source1",
]
`;

exports[`treeView reducer for the action DISPLAY_LEAF must update the visible leaves state and the expandedNodeIds for an album 1`] = `
Object {
  "album3": true,
  "album4": true,
  "album5": true,
  "source1": true,
}
`;

exports[`treeView reducer for the action DISPLAY_LEAF must update the visible leaves state and the expandedNodeIds for an album 2`] = `
Array [
  "source1",
  "album3",
  "album4",
  "album5",
]
`;

exports[`treeView reducer for the action EXPAND must display the correct leaf if it's a source 1`] = `
Object {
  "album1": false,
  "source1": true,
}
`;

exports[`treeView reducer for the action EXPAND must display the correct leaf if it's an album 1`] = `
Object {
  "album1": true,
  "album2": true,
  "source1": true,
}
`;

exports[`treeView reducer for the action TOGGLE must call collapse if it's visible 1`] = `
Object {
  "album1": false,
  "album2": true,
  "album3": true,
  "source1": true,
  "source2": true,
}
`;

exports[`treeView reducer for the action TOGGLE must call collapse if it's visible 2`] = `
Array [
  "source1",
  "album2",
  "album3",
  "source2",
]
`;

exports[`treeView reducer for the action TOGGLE must call expand if it's hidden 1`] = `
Object {
  "album1": true,
  "album2": false,
  "album3": true,
  "source1": true,
  "source2": true,
}
`;

exports[`treeView reducer for the action TOGGLE must call expand if it's hidden 2`] = `
Array [
  "album1",
  "source1",
  "album3",
  "source2",
]
`;

exports[`treeView reducer for the action UPDATE_EXPANDED_NODE_IDS must update the expandedNodeIds 1`] = `
Array [
  "album1",
  "source1",
  "album2",
  "album3",
  "source2",
]
`;

exports[`treeView reducer for the action UPDATE_EXPANDED_NODE_IDS must update the expandedNodeIds when everything is false 1`] = `Array []`;
